### PR TITLE
Add boilerplate config and code for gRPC and protobufs

### PR DESCRIPTION
**Pull request by dpezely**
This PR addresses first task of [issue #5](https://github.com/helium/helium-packet-router/issues/5) for Helium `packet-router` which functions as an intermediary gateway for LoRaWAN.

Notes: 

- This is **not a full router / LNS**
- For implementing the behavior, `helium_gateway_bhvr`, code was replicated from [sibyl](https://github.com/helium/sibyl) strictly for purposes of exercising the config and dependencies
   + Code was copied for merely attaining a clean compile via `make compile` (or simply, `make`)
   + Specifically, `helium-packet-router/src/grpc/helium_packet_router_service.erl` was replicated from `router`
   + This code **has NOT been executed**
- This is also to establish a baseline from which other branches can easily extend
   + Let's ensure that `rebar.config`, `sys.config`, etc. within this PR are minimal and clean
   + Let's keep rebar.config mostly alphabetical for its sequence of `plugins`, `prehooks`, etc.
- This is **without tests** of any kind but will be added in a future PR
   + because this PR is mostly about rebar.config, etc.
- Modules called from `src/grpc/helium_packet_router_service.erl` are omitted from this PR
  + Those will be in the next PR

**Updates:**

- The version of `.github/workflows/tests.yml` within `main` branch omits `make grpc` before `./rebar3 compile`, so this PR's build fails in CI
  + The fix is included here with commit 4ff36a4
  + But it apparently only gets applied when committed to `main` branch-- not the PR's own branch
  + *sigh*
- An early commit used same Erlang behavior from `sibyl` but changed to that from `router` for simplicity

@andymck I followed your trail of breadcrumbs through various repos such as `gateway-rs`, `blockchain-core`, `miner` and of course `sibyl`.  Thank you for all your earlier work on each of those.  If there's any legacy config or code that doesn't need to be ported here for `packet-router`, please let us know!

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -1,17 +1,38 @@
...
+
 {plugins, [
+    {grpcbox_plugin,
+        {git, "https://github.com/novalabsxyz/grpcbox_plugin.git", {branch, "master"}}},
     {rebar3_format, "1.2.1"},
     {erlfmt, "1.0.0"}
 ]}.
 
+{pre_hooks, [
+    {compile, "make grpc"},
```
Comment (michaeldjeffrey) [inline rebar.config:30]: Is there a way we can not have this pre_hook?
It recompiles grpc when you do anything with rebar3.

File: rebar.config [hunk 2]
```diff
@@ -1,17 +1,38 @@
-{deps, []}.
+{deps, [
+    {erlang_lorawan, ".*",
+        {git, "https://github.com/helium/erlang-lorawan.git", {branch, "master"}}},
+    {grpc_client,
+        {git, "https://github.com/Bluehouse-Technology/grpc_client.git", {branch, "master"}}},
+    {grpcbox, {git, "https://github.com/novalabsxyz/grpcbox.git", {branch, "master"}}},
+    {helium_proto, {git, "https://github.com/helium/proto.git", {branch, "master"}}},
```
Comment (ruttenberg) [inline rebar.config:7]: I don't think it is a good idea to specify a dependency via branch name. This practice leaves us open to accidentally absorbing library changes that weren't intended for release.
Yes, I know that the change won't be absorbed unless _rebar.lock_ is updated, but that has two weaknesses:

1.  The risk of accidentally committing the change is increased when other dependency changes are in _rebar.lock_.
2. Nothing is communicated about the  _intent_ of the change. When a tag (or at minimum, a commit hash) is used, it is clear that the change was consciously made.

We don't live in a world where mistakes are never committed to significant branches, nor where our testing schemes are likely to save us from not thinking about these changes.

File: rebar.config (hunk 1 above)
Comment (dpezely) [reply rebar.config]: Yes, there is a way.  It was already accounted for via `Makefile` such that this `pre_hook` was redundant.

e.g., within Makefile: `compile: | $(grpc_services_directory)`

Fixed in commit 5531bd9.

File: Makefile [hunk 3]
```diff
@@ -1,15 +1,17 @@
-.PHONY: compile clean test rel run docker-build docker-test docker-run
+.PHONY: compile clean test rel run grpc docker-build docker-test docker-run
+
+grpc_services_directory=src/grpc/autogen
 
 REBAR=./rebar3
 
-compile: |
-	$(REBAR) compile 
+compile: | $(grpc_services_directory)
```
Comment (andymck) [inline Makefile:7]: on miner we removed this pre requisite approach of handling grpc and instead used rebar prehooks.  Be worth checking that setup out.  Keeps things cleaner and more maintainable

File: src/grpc/helium_packet_router_service.erl [hunk 4]
```diff
@@ -0,0 +1,45 @@
+-module(helium_packet_router_service).
```
Comment (andymck) [inline src/grpc/helium_packet_router_service.erl:1]: noting it here for reference, this handler was one of the first grpc handlers added and its approach  with the 'receive' loop is maybe a little raw.

The newer handlers implement handle_info msgs and as such are a somewhat more of an otp style approach.  An example of those newer handlers can be seen here:

https://github.com/helium/blockchain-core/blob/master/src/grpc/blockchain_grpc_sc_server_handler.erl

This router handler is obviously a lot simpler and like i say nothing wrong with it on its own but maybe be better to follow the newer pattens of the later handlers especially if this is going to be the basis of other handlers

File: config/grpc_client_gen.config [hunk 5]
```diff
@@ -0,0 +1,34 @@
+{plugins, [
```
Comment (andymck) [inline config/grpc_client_gen.config:1]: note: client config only required if you plan on adding grpc tests and need to execute client behaviour

File: rebar.config [hunk 6]
```diff
@@ -1,17 +1,36 @@
...
+    {lager, "3.9.2"}
+]}.
 
 {erl_opts, [
     debug_info,
+    {parse_transform, lager_transform},
     warnings_as_errors
 ]}.
 
+{grpc, [
```
Comment (andymck) [inline rebar.config:19]: i dont think you need this section

File: rebar.config [hunk 7]
```diff
@@ -1,17 +1,36 @@
...
+    {gpb_opts, [{module_name_suffix, "_pb"}]}
+]}.
+
 {plugins, [
+    {grpcbox_plugin, {git, "https://github.com/novalabsxyz/grpcbox_plugin.git"}},
     {rebar3_format, "1.2.1"},
     {erlfmt, "1.0.0"}
 ]}.
 
+{pre_hooks, []}.
```
Comment (andymck) [inline rebar.config:30]: as per previous comment, better to call the grpc makefile target here from a compile pre hook.  likewise for a clean prehook.  see https://github.com/helium/miner/blob/master/rebar.config#L61

File: Makefile (hunk 3 above)
Comment (dpezely) [reply Makefile:7]: Apparently, one side-effect of using rebar3's `pre_hooks` is that grpc modules get regenerated too frequently.

We originally had `pre_hooks` consistent with miner in the first commit of this branch but requested to be removed from `rebar.conf`, so then `make compile` for full builds and `rebar compile` for rapid iterations.

Is there a flaw in that reasoning?

File: config/grpc_client_gen.config (hunk 5 above)
Comment (dpezely) [reply config/grpc_client_gen.config]: Removed with commit c300421

File: rebar.config (hunk 6 above)
Comment (dpezely) [reply rebar.config]: Removed with commit 2181801

Confirmed: `make clean ; rebar3 grpc gen` still works

File: src/grpc/helium_packet_router_service.erl (hunk 4 above)
Comment (macpie) [reply src/grpc/helium_packet_router_service.erl:1]: ok I think we will be able to use this, this is just boilerplate to start including GRPCwe actually won't even use this service in there

File: rebar.config [hunk 8]
```diff
@@ -1,5 +1,12 @@
+{cover_export_enabled, true}.
+
+{cover_enabled, true}.
+
 {deps, [
-    {lager, "3.9.2"}
+    {grpcbox, {git, "https://github.com/novalabsxyz/grpcbox.git"}},
+    {helium_proto, {git, "https://github.com/helium/proto.git"}},
+    {lager, "3.9.2"},
+    {lorawan, {git, "https://github.com/helium/erlang-lorawan.git"}}
```
Comment (macpie) [inline rebar.config:9]: I don't think we need this one in there I ll be adding it as part of another PR

File: config/sys.config [hunk 9]
```diff
@@ -1,4 +1,30 @@
 [
+    {grpcbox, [
```
Comment (macpie) [inline config/sys.config:2]: we should include this in `sys.config.src` as well as it will be used for release

Comment (dpezely) [reply config/sys.config:2]: Thanks!  I keep forgetting about `sys.conf.src`...

Fixed in commit 3884fc8

File: rebar.config (hunk 8 above)
Comment (dpezely) [reply rebar.config]: Removed dependency.  Fixed in commit 3884fc8

File: rebar.config (hunk 7 above)
Comment (dpezely) [reply rebar.config:23]: (as above: let's merge with the current balance of Makefile vs rebar.config and revisit in a future PR if needed)

**Review by macpie — APPROVED** [review]
